File: ios/PocketSensorKit/Sources/PocketSensorCore/Conversion/Units.swift
import Foundation
import simd

/// Apple の単位から wire の SI 単位への変換。計算は Double。
public enum Units {
    public static let standardGravity = 9.80665

    /// 深度の無効判定。65.535 m 以上は uint16 の mm に収まらない。
    public static let depthMaxM = 65.535

    public static func accelGToMps2(_ g: SIMD3<Double>) -> SIMD3<Double> {
        g * -standardGravity
    }

    public static func magMicroTeslaToTesla(_ microTesla: SIMD3<Double>) -> SIMD3<Double> {
        microTesla * 1e-6
    }

    public static func pressureKPaToPa(_ kPa: Double) -> Double {
        kPa * 1000.0
    }

    /// (-π, π] へ畳む。-π は π にする。
    public static func wrapToPi(_ angle: Double) -> Double {
        let twoPi = 2.0 * Double.pi
        var x = angle.truncatingRemainder(dividingBy: twoPi)
        if x > Double.pi {
            x -= twoPi
        } else if x <= -Double.pi {
            x += twoPi
        }
        return x
    }

    /// 真北 0° 時計回りを、東 0 rad 反時計回りへ。
    public static func courseDegToENUYaw(_ courseDeg: Double) -> Double {
        wrapToPi(.pi / 2.0 - courseDeg * .pi / 180.0)
    }

    public static func accuracyToVariance(_ accuracy: Double) -> Double {
        accuracy * accuracy
    }

    /// 水平精度が負なら NO_FIX。垂直が負なら Up の分散だけ 0。
    public static func navSatCovariance(
        horizontalAccuracy: Double,
        verticalAccuracy: Double
    ) -> (covariance: [Double], type: UInt8, status: Int8) {
        if horizontalAccuracy < 0.0 {
            return (
                Array(repeating: 0.0, count: 9),
                SensorMsgs.NavSatFix.covarianceTypeUnknown,
                SensorMsgs.NavSatStatus.statusNoFix
            )
        }
        let hh = accuracyToVariance(horizontalAccuracy)
        let vv = verticalAccuracy < 0.0 ? 0.0 : accuracyToVariance(verticalAccuracy)
        var cov = Array(repeating: 0.0, count: 9)
        cov[0] = hh
        cov[4] = hh
        cov[8] = vv
        return (cov, SensorMsgs.NavSatFix.covarianceTypeApproximated, SensorMsgs.NavSatStatus.statusFix)
    }

    /// 画素ごとに m を mm の uint16 へ。NaN、0 以下、65.535 以上は 0。半上げは Float を Double にして行う。
    public static func depthMetersToMillimeters(
        base: UnsafeRawPointer,
        width: Int,
        height: Int,
        bytesPerRow: Int
    ) -> Data {
        var out = Data(count: width * height * MemoryLayout<UInt16>.size)
        out.withUnsafeMutableBytes { dest in
            let dst = dest.bindMemory(to: UInt16.self)
            var i = 0
            for row in 0 ..< height {
                let rowBase = base.advanced(by: row * bytesPerRow)
                for col in 0 ..< width {
                    let d = Double(rowBase.load(fromByteOffset: col * MemoryLayout<Float>.size, as: Float.self))
                    dst[i] = depthSampleToMillimeters(d)
                    i += 1
                }
            }
        }
        return out
    }

    public static func depthMetersToMillimeters(_ pixels: [Float]) -> Data {
        pixels.withUnsafeBufferPointer { buf in
            guard let base = buf.baseAddress else { return Data() }
            return depthMetersToMillimeters(
                base: UnsafeRawPointer(base),
                width: pixels.count,
                height: 1,
                bytesPerRow: pixels.count * MemoryLayout<Float>.size
            )
        }
    }

    /// 行末の詰めを除いて詰める。confidence の uint8 画像に使う。
    public static func packRows(
        base: UnsafeRawPointer,
        width: Int,
        height: Int,
        bytesPerRow: Int,
        bytesPerPixel: Int
    ) -> Data {
        let rowBytes = width * bytesPerPixel
        var out = Data(count: rowBytes * height)
        out.withUnsafeMutableBytes { dest in
            guard let dst = dest.baseAddress else { return }
            for row in 0 ..< height {
                dst.advanced(by: row * rowBytes).copyMemory(
                    from: base.advanced(by: row * bytesPerRow),
                    byteCount: rowBytes
                )
            }
        }
        return out
    }

    private static func depthSampleToMillimeters(_ d: Double) -> UInt16 {
        if d.isNaN || d <= 0.0 || d >= depthMaxM {
            return 0
        }
        return UInt16(floor(d * 1000.0 + 0.5))
    }
}

File: python/src/pocketsensor/units.py
"""Apple の単位から wire の SI 単位への変換。計算は float64。"""

from __future__ import annotations

import math
from typing import Final

import numpy as np
from numpy.typing import ArrayLike, NDArray

STANDARD_GRAVITY: Final[float] = 9.80665

# sensor_msgs/NavSatStatus と NavSatFix の定数
STATUS_NO_FIX: Final[int] = -1
STATUS_FIX: Final[int] = 0
COVARIANCE_TYPE_UNKNOWN: Final[int] = 0
COVARIANCE_TYPE_APPROXIMATED: Final[int] = 1

# 深度の無効判定。65.535 m 以上は uint16 の mm に収まらない
_DEPTH_MAX_M: Final[float] = 65.535


def accel_g_to_mps2(v3: ArrayLike) -> NDArray[np.float64]:
    """比力へ直す。Core Motion の符号を反転し、標準重力を掛ける。"""
    return np.asarray(v3, dtype=np.float64) * -STANDARD_GRAVITY


def mag_ut_to_tesla(v3: ArrayLike) -> NDArray[np.float64]:
    return np.asarray(v3, dtype=np.float64) * 1e-6


def pressure_kpa_to_pa(x: float) -> float:
    return float(np.float64(x) * 1000.0)


def wrap_to_pi(angle: float) -> float:
    """(-pi, pi] へ畳む。-pi は pi にする。"""
    two_pi = 2.0 * math.pi
    x = math.fmod(float(angle), two_pi)
    if x > math.pi:
        x -= two_pi
    elif x <= -math.pi:
        x += two_pi
    return x


def course_deg_to_enu_yaw(course_deg: float) -> float:
    """真北 0° 時計回りを、東 0 rad 反時計回りへ。"""
    return wrap_to_pi(math.pi / 2.0 - math.radians(float(course_deg)))


def accuracy_to_variance(a: float) -> float:
    v = np.float64(a)
    return float(v * v)


def navsat_covariance(
    horizontal_accuracy_m: float,
    vertical_accuracy_m: float,
) -> tuple[list[float], int, int]:
    """水平精度が負なら NO_FIX。垂直が負なら Up の分散だけ 0。"""
    h = float(horizontal_accuracy_m)
    v = float(vertical_accuracy_m)
    if h < 0.0:
        return [0.0] * 9, COVARIANCE_TYPE_UNKNOWN, STATUS_NO_FIX
    hh = accuracy_to_variance(h)
    vv = 0.0 if v < 0.0 else accuracy_to_variance(v)
    cov = [0.0] * 9
    cov[0] = hh
    cov[4] = hh
    cov[8] = vv
    return cov, COVARIANCE_TYPE_APPROXIMATED, STATUS_FIX


def depth_m_to_mm_u16(depth: ArrayLike) -> NDArray[np.uint16]:
    """画素ごとに m を mm の uint16 へ。NaN、0 以下、65.535 以上は 0。半上げは float64 で行う。"""
    d = np.asarray(depth, dtype=np.float64)
    out = np.zeros(d.shape, dtype=np.uint16)
    valid = ~(np.isnan(d) | (d <= 0.0) | (d >= _DEPTH_MAX_M))
    if np.any(valid):
        out[valid] = np.floor(d[valid] * 1000.0 + 0.5).astype(np.uint16)
    return out
